server/src/updaters/thread-updaters.js

```diff
@@ -427,30 +429,65 @@ async function updateThread(
     nextThreadType !== oldThreadType ||
     nextParentThreadID !== oldParentThreadID
   ) {
-    savePromises.recalculatePermissionsChangeset = recalculateAllPermissions(
+    intermediatePromises.recalculatePermissionsChangeset = recalculateAllPermissions(
       request.threadID,
       nextThreadType,
     );
   }
   const {
     addMembersChangeset,
     recalculatePermissionsChangeset,
-  } = await promiseAll(savePromises);
+  } = await promiseAll(intermediatePromises);
 
-  const changeset = [];
+  const membershipRows = [];
+  const relationshipRows = [];
   if (recalculatePermissionsChangeset && newMemberIDs) {
-    changeset.push(
-      ...recalculatePermissionsChangeset.filter(
+    const {
+      membershipRows: recalculateMembershipRows,
+      relationshipRows: recalculateRelationshipRows,
+    } = recalculatePermissionsChangeset;
+    membershipRows.push(
+      ...recalculateMembershipRows.filter(
         rowToSave => !newMemberIDs.includes(rowToSave.userID),
       ),
     );
+    const parentMemberIDs = recalculateMembershipRows
+      .filter(rowToSave => !newMemberIDs.includes(rowToSave.userID))
+      .map(rowToSave => rowToSave.userID);
+    const newUserIDs = newMemberIDs.filter(
+      memberID =>
+        !recalculateMembershipRows.find(
+          rowToSave => memberID === rowToSave.userID,
+        ),
+    );
+    const parentRelationshipRows = cartesianProduct(
+      parentMemberIDs,
+      newUserIDs,
+    ).map(pair => {
+      const [user1, user2] = sortIDs(...pair);
+      const status = undirectedStatus.KNOW_OF;
+      return { user1, user2, status };
+    });
+    relationshipRows.push(
+      ...recalculateRelationshipRows,
+      ...parentRelationshipRows,
+    );
```

Comment:
This code is very similar to what we do in `createThread`. Can you factor it out so we can share it? I'm thinking of a function `getParentThreadMembershipRowsForNewUsers` that takes `recalculateMembershipRows` and `newMemberIDs`